grpc/src/main/java/com/linecorp/armeria/server/grpc/ProtobufRequestConverterFunction.java

```diff
@@ -0,0 +1,160 @@
+        if (Message.class.isAssignableFrom(expectedResultType)) {
+            final Message.Builder messageBuilder = getMessageBuilder(expectedResultType);
+
+            if (contentType == null ||
+                contentType.subtype().contains("protobuf") || contentType.is(MediaType.OCTET_STREAM)) {
+                return messageBuilder.mergeFrom(request.content().array(), extensionRegistry).build();
+            }
+
+            if (isJson(contentType)) {
+                jsonParser.merge(request.contentUtf8(), messageBuilder);
+                return messageBuilder.build();
+            }
+        }
+
+        if (isJson(contentType) && expectedParameterizedResultType != null) {
+            final int collectionType;
+            if (List.class.isAssignableFrom(expectedResultType)) {
+                collectionType = 1;
+            } else if (Set.class.isAssignableFrom(expectedResultType)) {
+                collectionType = 2;
+            } else {
+                collectionType = 0;
+            }
+
+            if (collectionType > 0) {
+                final Class<?> typeArgument =
+                        (Class<?>) expectedParameterizedResultType.getActualTypeArguments()[0];
+                if (Message.class.isAssignableFrom(typeArgument)) {
+                    final String content = request.content(contentType.charset(StandardCharsets.UTF_8));
+
+                    final JsonNode jsonNode = mapper.readTree(content);
+                    if (jsonNode.isArray()) {
+                        final ImmutableCollection.Builder<Message> builder;
+                        if (collectionType == 1) {
+                            builder = ImmutableList.builderWithExpectedSize(jsonNode.size());
+                        } else {
+                            builder = ImmutableSet.builderWithExpectedSize(jsonNode.size());
+                        }
+                        for (JsonNode node : jsonNode) {
+                            final Message.Builder messageBuilder = getMessageBuilder(typeArgument);
+                            jsonParser.merge(mapper.writeValueAsString(node), messageBuilder);
+                            builder.add(messageBuilder.build());
+                        }
+
+                        return builder.build();
+                    }
+                }
+            }
+        }
+
+        return RequestConverterFunction.fallthrough();
+    }
+
+    private static boolean isJson(@Nullable MediaType contentType) {
+        return contentType != null &&
+               (contentType.is(MediaType.JSON) || contentType.subtype().endsWith("+json"));
+    }
+
+    private static Message.Builder getMessageBuilder(Class<?> clazz) throws Exception {
+        Method method = methodCache.get(clazz);
+        if (method == null) {
+            method = clazz.getMethod("newBuilder");
+            methodCache.put(clazz, method);
+        }
+        return (Message.Builder) method.invoke(clazz);
```

Comment:
I was just wondering how slower is `Method.invoke(...)` than `MethodHandle.invoke(...)`.  
```
Benchmark                          Mode  Cnt         Score         Error  Units
MethodCallBenchmark.directCall    thrpt    5  30697469.267 ± 5620688.880  ops/s
MethodCallBenchmark.method        thrpt    5  24412855.684 ± 4318917.364  ops/s
MethodCallBenchmark.methodHandle  thrpt    5  29439498.151 ± 1840979.485  ops/s
```

There is 20% performance degression in `Method.invoke(...)`. Let me change this with `MethodHandle`. :-)